### PR TITLE
Rec: three more coverity cases: 1487743, 1419403 and 1419401

### PR DESCRIPTION
**Pull request by omoerbeek**
### Short description

With these three we no longer have outstanding coverity issues.

Check the commit message of the 2nd commit for some background info.

Once this is in, I plan to
- see if adding a model would reduce the amount of false positives. In particular, coverity has some issues understanding passing pointer values through pipes and sees those as leaks.
- review closed reports to see if we might have dismissed reports wrongly

### Checklist

I have:
- [X] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [X] compiled this code
- [X] tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

**Comment by rgacogne** [comment]
> see if adding a model would reduce the amount of false positives. In particular, coverity has some issues understanding passing pointer values through pipes and sees those as leaks.

I believe #11427 might help with that, at least by reducing the number of places a pointer is written to a pipe to a single one. It still requires a fair amount of work, though.

File: pdns/iputils.hh [hunk 1]
```diff
@@ -884,17 +884,20 @@ private:
...
+      if (node != nullptr)
+        node = node->traverse_l();
+      while (node != nullptr) {
+        if (node->assigned)
+          insert(node->node.first).second = node->node.second;
+        node = node->traverse_lnr();
+      }
+    }
+    catch (...) {
+      abort();
```
Comment (rgacogne) [inline pdns/iputils.hh:900]: I agree that copying an existing tree is probably safe in most cases, but calling `abort()` if we run out of memory while trying to copy a large NetmaskTree seems wrong?

Comment (omoerbeek) [reply pdns/iputils.hh:900]: I  am not very happy with that either, reading up on the exact semantics of `noexcept' to see if we can only catch `NetmaskException` here.

Comment (omoerbeek) [reply pdns/iputils.hh:900]: If I understand correctly, `noexcept` fucntions that actually throw cause a runtime kill.
So `noexcept` cannot be used if we cacth `NetmaskException` only, as allocation failures might still throw. Calling code would have a chance to react on OOM situations. Let's see if Coverity likes that enough.

File: pdns/iputils.hh [hunk 2]
```diff
@@ -964,14 +967,14 @@ public:
...
-  NetmaskTree() noexcept: d_root(new TreeNode()), d_left(nullptr), d_size(0) {
+  NetmaskTree(): d_root(new TreeNode()), d_left(nullptr), d_size(0) {
   }
 
   NetmaskTree(const NetmaskTree& rhs): d_root(new TreeNode()), d_left(nullptr), d_size(0) {
     copyTree(rhs);
   }
 
-  NetmaskTree& operator=(const NetmaskTree& rhs) {
+  NetmaskTree& operator=(const NetmaskTree& rhs) noexcept {
```
Comment (rgacogne) [inline pdns/iputils.hh:977]:
```suggestion
  NetmaskTree& operator=(const NetmaskTree& rhs) {
```

**Review by rgacogne — CHANGES_REQUESTED** [review]

**Review by rgacogne — APPROVED** [review]
:+1:

**Comment by rgacogne** [comment]
> yes, in the meantime, I'm wondering why an dynamic allocation is done while the write and read seem to be blocking . We could just allocate on the stack, pass the pointer and worry no more...
This is from looking at the code, did not test it yet, might overlook something.

In most cases I can think of, especially in dnsdist, that would not work because we are often going to queue several objects into the pipe before the other end even start reading. Even if the other cases that sounds very brittle, what happens if there is  a timeout and the sender has exited the function when the receiver reads from the pipe, for example?

**Comment by omoerbeek** [comment]
> > yes, in the meantime, I'm wondering why an dynamic allocation is done while the write and read seem to be blocking . We could just allocate on the stack, pass the pointer and worry no more...
> > This is from looking at the code, did not test it yet, might overlook something.
> 
> In most cases I can think of, especially in dnsdist, that would not work because we are often going to queue several objects into the pipe before the other end even start reading. Even if the other cases that sounds very brittle, what happens if there is a timeout and the sender has exited the function when the receiver reads from the pipe, for example?

Agreed, this was a bad idea as I already found out myself as well.

**Pull request merged**